WIP: Add YAML tests for conflicts during dependency resolution

ilanschnell opened the pull request:
This PR adds some conflicting yaml tests.  The purpose of these tests is to test the error messages of the resolver.

pradyunsg commented:
(CI failed due to networking issues; we can trigger re-runs if stuff still fails later)

pradyunsg commented:
@ilanschnell How close is this to being ready-to-merge?

Also, FYI: `test_yaml_based[large-2*old]` fails on PyPy.

ilanschnell commented:
@pradyunsg @pfmoore @uranusjr How detailed should the testing for error messages be?  It would be obviously possible to just hard code the entire expected error message into the yaml file.  However, this would mean that even the smallest change would break the test.  I think it would be better to check for specific words that appear in the error message, or do maybe do some regular expression matching.

uranusjr commented:
The new resolver has some problems with the current suite being too pedantic about having the exact same message, but the current message format simply does not make sense for the backtracking resolver. I would prefer to only test some keywords are in the message so it allows elbow room for pip to potentially improve in some areas.

pfmoore commented:
Can we do a regex match? That's probably sufficiently general to cover most bases. It may be more than we need, but it reduces the risk that we need to rewrite the engine for an unexpected case - e.g., case insensitivity can be handled with the `(?i)` pattern.

pradyunsg commented:
Making them a regex match, sounds like a good plan to me!

ilanschnell commented:
Thanks for the feedback.  I like the regex approach as it seems a natural extension to simple string matching.  I'll try to implement this is a flexible way in order to allow future cases.

@pradyunsg It's very strange that only that one test is failing under PyPy.

pradyunsg commented:
@uranusjr I'd avoided adding that label so that I don't accidentally merge this prematurely. :)

brainwane commented:
>  I'll try to implement this is a flexible way in order to allow future cases.

@ilanschnell Will you finish this today?

pfmoore commented:
I haven't done a full review here, but if we can get error message checking via regex, this looks like it would be in a good place.

ilanschnell commented:
@brainwane Yes, I'll finish this today

ilanschnell commented:
@pradyunsg The PR is ready for review and merging now.

brainwane commented:
I alerted @ilanschnell to Pradyun's review and he'll wrap up his response to that review tomorrow. If he isn't able to get it to a mergeable state by tomorrow then we'll have another developer take it over from here.

ilanschnell commented on tests/functional/test_yaml.py:
```diff
@@ -175,24 +185,17 @@ def test_yaml_based(script, case):
+            assert effect['state'] == (response['state'] or []), str(result)
+
+        error = response.get('error')
+        if error and case[":resolver:"] == 'new' and sys.platform != 'win32':
```
I skipped this on Windows, because it was giving me problems, but then I forgot looking into it in more detail later.  Exit codes and stderr shouldn't be treated differently on Windows, so I'm not sure why it was failing.

ilanschnell commented on tests/yaml/conflict_1.yml:
```diff
@@ -0,0 +1,78 @@
+    - state: null
+      error:
+        code: 1
+        stderr: 'no\s+matching\s+distribution'
```
Yes, the reason it put `\s+` here was to avoid formatting changes in the future.  I agree that `' no matching distribution'` is cleaner to read :-)

ilanschnell commented on tests/functional/test_yaml.py:
```diff
@@ -137,6 +124,29 @@ def handle_request(script, action, requirement, options, new_resolver=False):
+
+    for patter in patters:
+        pat = re.compile(patter, re.I)
+        match = pat.search(result.stderr)
```
Ah, the `re.I` flag got dropped, which is why some tests stared failing.

ilanschnell commented:
Thanks for the feedback @pradyunsg.  I found a small problem with your regex change (the `re.I` flag got lost), which I fixed.  Also I added a comment on why the errer checking test on Windows is skipped.

uranusjr commented:
I looked into the PyPy failure, and it seems like PyPy has additional built-in modules to CPython (namely `cffi`), which makes the package not being installed. I’m still thinking how is best to deal with this. Maybe we should just rewrite the YAML declaration to avoid those packages, or maybe we should implement additional logic to identify those packages as “installed” on PyPy.

pradyunsg commented:
I would just change `cffi` to `magic-cffi` and add a comment about why we don't use the package name `cffi`. :)
